### ISSUE TITLE
aks failed to get mounted files, err: failed to list all files in target path

### ISSUE THREAD

**Issue by jrudley** [Azure/secrets-store-csi-driver-provider-azure#268]
**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
I am only seeing this error on k8s init windows jobs. I am unsure of linux jobs are impacted as I don't run any linux init jobs.
```
13m         Warning   FailedToCreateSecret   /mssql-init-sr859   failed to get mounted files, err: failed to list all files in target path c:\var\lib\kubelet\pods\a7529eb1-16eb-4b23-9674-d8fc2c7e8424\volumes\kubernetes.io~csi\secrets-store-inline\mount, err: open c:\var\lib\kubelet\pods\a7529eb1-16eb-4b23-9674-d8fc2c7e8424\volumes\kubernetes.io~csi\secrets-store-inline\mount: The system cannot find 
the path specified.
10m         Warning   FailedToCreateSecret   /solr-init-pgzvc    failed to get mounted files, err: failed to list all files in target path c:\var\lib\kubelet\pods\bd78698c-b15e-4d71-bf6d-e58f4e125f0e\volumes\kubernetes.io~csi\secrets-store-inline\mount, err: open c:\var\lib\kubelet\pods\bd78698c-b15e-4d71-bf6d-e58f4e125f0e\volumes\kubernetes.io~csi\secrets-store-inline\mount: The system cannot find 
the path specified.

```
This does not stop the k8s job from running and completing successfully.

I did not see this issue in 0.0.8, but I am on 0.0.9
aks 1.18.8 vmss with latest windows node image

**What did you expect to happen:**
no warning error.

**Which access mode did you use to access the Azure Key Vault instance:**
[e.g. Service Principal, Pod Identity, User Assigned Managed Identity, System Assigned Managed Identity]
User Assigned Managed Identity

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): 0.0.16
- Azure Key Vault provider version: (use the image tag):0.0.9
- Kubernetes version: (use `kubectl version` and `kubectl get nodes -o wide`):
```
kubectl version

Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"73ec19bdfc6008cd3ce6de96c663f70a69e2b8fc", GitTreeState:"clean", BuildDate:"2020-09-17T04:17:08Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}

kubectl get nodes -o wide

NAME                                STATUS   ROLES   AGE   VERSION   INTERNAL-IP     EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME
aks-nodepool1-27591730-vmss000000   Ready    agent   8h    v1.18.8   10.200.12.4     <none>        Ubuntu 18.04.5 LTS               5.4.0-1026-azure   docker://19.3.12
aks-nodepool1-27591730-vmss000001   Ready    agent   8h    v1.18.8   10.200.12.20    <none>        Ubuntu 18.04.5 LTS               5.4.0-1026-azure   docker://19.3.12
aks-npsolr-27591730-vmss000000      Ready    agent   8h    v1.18.8   10.200.12.129   <none>        Ubuntu 18.04.5 LTS               5.4.0-1026-azure   docker://19.3.12
aks-npsolr-27591730-vmss000001      Ready    agent   8h    v1.18.8   10.200.12.140   <none>        Ubuntu 18.04.5 LTS               5.4.0-1026-azure   docker://19.3.12
aks-npsolr-27591730-vmss000002      Ready    agent   8h    v1.18.8   10.200.12.151   <none>        Ubuntu 18.04.5 LTS               5.4.0-1026-azure   docker://19.3.12
aks-npzk-27591730-vmss000000        Ready    agent   8h    v1.18.8   10.200.12.162   <none>        Ubuntu 18.04.5 LTS               5.4.0-1026-azure   docker://19.3.12
aks-npzk-27591730-vmss000001        Ready    agent   8h    v1.18.8   10.200.12.173   <none>        Ubuntu 18.04.5 LTS               5.4.0-1026-azure   docker://19.3.12
aks-npzk-27591730-vmss000002        Ready    agent   8h    v1.18.8   10.200.12.184   <none>        Ubuntu 18.04.5 LTS               5.4.0-1026-azure   docker://19.3.12
aksnpwin000000                      Ready    agent   8h    v1.18.8   10.200.12.36    <none>        Windows Server 2019 Datacenter   10.0.17763.1397    docker://19.3.11
aksnpwin000001                      Ready    agent   8h    v1.18.8   10.200.12.67    <none>        Windows Server 2019 Datacenter   10.0.17763.1397    docker://19.3.11
aksnpwin000002                      Ready    agent   8h    v1.18.8   10.200.12.98    <none>        Windows Server 2019 Datacenter   10.0.17763.1397    docker://19.3.11

```
- Cluster type: (e.g. AKS, aks-engine, etc): AKS

**Comment by aramase** [comment]
@jrudley Could you provide the following -
1. `SecretProviderClass` YAML
2. Job YAML

`v0.0.16` generates events in case of failures. It's possible these errors existed in the logs before but wasn't concerning?

**Comment by jrudley** [comment]
```
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: azure-sync-solr
spec:
  provider: azure
  secretObjects:                           # [OPTIONAL] SecretObject defines the desired state of synced K8s secret objects
  - secretName: solr-certificate-secret
    type: Opaque
    data: 
    - objectName: "solr-certificate-secret"                   # name of the mounted content to sync. this could be the object name or object alias 
      key: solr-certificate-secret
  - secretName: solr-bastion
    type: kubernetes.io/tls
    data: 
    - objectName: solr-bastion
      key: tls.key
    - objectName: solr-bastion
      key: tls.crt 
  parameters:
    useVMManagedIdentity: "true"
    userAssignedIdentityID: "<identityid>"
    keyvaultName: "<kvname>"               # the name of the KeyVault
    objects: |
      array:
        - |
          objectName: solr-certificate-secret
          objectType: secret       # object types: secret, key or cert
        - |
          objectName: solr-certificate
          objectType: secret #use secret to get public and private key. cert only gets public. key gets private
        - |
          objectName: solr-bastion
          objectType: secret
    tenantId: "<tenantid>" # the tenant ID of the KeyVault
```

```
apiVersion: batch/v1
kind: Job
metadata:
  name: solr-init
spec:
  template:
    spec:
      nodeSelector:
        kubernetes.io/os: windows 
      volumes:
        - name: secrets-store-inline
          csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: "azure-sync" 
      imagePullSecrets:
        - name: sitecore-docker-registry
      containers:
      - name: solr-init
        image: <acrname>.azurecr.io/solr-init-xp:latest
        volumeMounts:
        - name: secrets-store-inline 
          mountPath: "/mnt/secrets-store"
          readOnly: true
        env:
        - name: SITECORE_SOLR_CONNECTION_STRING
          valueFrom:
            secretKeyRef:
              name: sitecore-solr-connection-string
              key: sitecore-solr-connection-string
        - name: SOLR_REPLICATION_FACTOR
          value: "3"
        - name: SOLR_CERT_NAME
          value: "solr-certificate"
        - name: SOLR_TLS
          value: "true"
      restartPolicy: Never
      tolerations:
        - key: "os"
          operator: "Equal"
          value: "windows"
          effect: "NoSchedule"
  backoffLimit: 5
```

Let me run some tests today and will report back my findings to give some more info

**Comment by jrudley** [comment]
I reprovisioned my cluster and cannot repo this now. I did use a custom name space in my new deploys, so maybe that had something to do with it?

**Comment by jrudley** [comment]
Reopening..I am seeing this on multiple pods on different windows nodes. When I apply a new yaml is when I see this error.
```
  Warning  FailedToCreateSecret  52s (x16 over 3m43s)  csi-secrets-store-controller  failed to get mounted files, err: failed to list all files in target path c:\var\lib\kubelet\pods\46b2da65-604d-43d4-94d4-5adb1b6eae26\volumes\kubernetes.io~csi\secrets-store-inline\mount, err: open c:\var\lib\kubelet\pods\46b2da65-604d-43d4-94d4-5adb1b6eae26\volumes\kubernetes.io~csi\secrets-store-inline\mount: The system cannot find the path specified.
```

This time, it isn't a job like I initially mentioned. Happens on any of my windows pods.
k logs azure-csi-secrets-store-csi-driver-windows-k5kvl secrets-store --tail=100
```
time="2020-10-20T23:11:39Z" level=debug msg="patcher completed"
time="2020-10-20T23:11:42Z" level=info msg="reconcile started" node=aksnpwin000001 secretproviderclasspodstatus=nonprod1/cm-97589bddf-wpmtt-nonprod1-azure-sync
time="2020-10-20T23:11:42Z" level=error msg="failed to get mounted files, err: failed to list all files in target path c:\\var\\lib\\kubelet\\pods\\27a24af7-73e4-4caf-a37b-a19974d8aa72\\volumes\\kubernetes.io~csi\\secrets-store-inline\\mount, err: open c:\\var\\lib\\kubelet\\pods\\27a24af7-73e4-4caf-a37b-a19974d8aa72\\volumes\\kubernetes.io~csi\\secrets-store-inline\\mount: The system cannot find the path specified." node=aksnpwin000001 secretproviderclasspodstatus=nonprod1/cm-97589bddf-wpmtt-nonprod1-azure-sync
time="2020-10-20T23:11:44Z" level=debug msg="GRPC call: /csi.v1.Identity/Probe"
```

I have to forcefully delete the pod.

**Comment by aramase** [comment]
@jrudley Are those pods using the `secretObjects`. `FailedToCreateSecret` indicates there are `secretObjects` in `SecretProviderClass`. The reconciler tries periodically to create the secret if initial attempt failed. 

So in this scenario, the pod is being deleted and the SecretProviderClass still exists?

**Comment by jrudley** [comment]
@aramase The one secretproviderclass named azure-sync is being used by multiple pods. When one of the pods that is using the azyre-sync secretproviderclass goes to terminate from a deployment, it seems that maybe the reconciler is hitting and halting the termination of the pod. Hopefully I am understanding your question correctly?

**Comment by aramase** [comment]
@jrudley Thank you for the information. The reconciler will only generate an event for the pod but shouldn't prevent it from terminating. I'll take a look.

**Comment by jrudley** [comment]
@aramase Thanks. Let me know how I can help.

**Comment by hli07** [comment]
Hi, I am getting the error: failed to get mounted files, err: failed to list all files in target path as well, though the situation is a little bit different how it ends up. And I am on linux image.

Access mode: SPN
Secrets Store CSI Driver version: 0.0.13
Azure Key Vault provider version: 0.0.9

`3m34s       Warning   FailedToCreateSecret   /inchcape-platform-api-dev-job-multi-dimensions-62h84             failed to get mounted files, err: failed to list all files in target path /var/lib/kubelet/pods/e08374d9-1f45-4f95-96ed-5ca6e975f397/volumes/kubernetes.io~csi/secrets-store/mount, err: open /var/lib/kubelet/pods/e08374d9-1f45-4f95-96ed-5ca6e975f397/volumes/kubernetes.io~csi/secrets-store/mount: no such file or directory
9m25s       Warning   FailedToCreateSecret   /inchcape-platform-api-dev-job-setup-project-database-rt5xc       failed to get mounted files, err: failed to list all files in target path /var/lib/kubelet/pods/8522bd7b-19c4-4083-874c-a3b1163f69ea/volumes/kubernetes.io~csi/secrets-store/mount, err: open /var/lib/kubelet/pods/8522bd7b-19c4-4083-874c-a3b1163f69ea/volumes/kubernetes.io~csi/secrets-store/mount: no such file or directory`

I found out some secrets didn't get populated in SecretProviderClass though I am very sure how it was sourced. The problem is gone after purging the deployment and doing a redeployment, @aramase any clue why secrets aren't updated from SecretProviderClass?

**Comment by aramase** [comment]
> I found out some secrets didn't get populated in SecretProviderClass though I am very sure how it was sourced. The problem is gone after purging the deployment and doing a redeployment, @aramase any clue why secrets aren't updated from SecretProviderClass?

@hli07 Are you using driver version `v0.0.13`? If the `SecretProviderClass` is updated after the pod is deployed then the pod needs to be restarted to pick up the new objects added to `SecretProviderClass`. The volume is mounted and populated only at the pod start. 

In driver version `v0.0.16` we have added support for auto rotation. This is in alpha feature and here are some useful links on how to enable it -
https://github.com/Azure/secrets-store-csi-driver-provider-azure#optional-enable-auto-rotation-of-secrets
https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/docs/README.rotation.md

**Comment by hli07** [comment]
`@hli07 Are you using driver version v0.0.13?` 
yes, should I update it?

`If the SecretProviderClass is updated after the pod is deployed then the pod needs to be restarted to pick up the new objects added to SecretProviderClass. The volume is mounted and populated only at the pod start.` 
I didn't manually update SecretProviderClass, It's part of my helm chart deployment which I just ensure SecretProviderClass gets updated with new values and that get passed it via the helm values before the helm deployment. Meaning it also redeploys all the jobs/pods(deploymeny kind)/cron etc, so the pods are restarted but it isn't syncing the latest secret that was added which I can see it from listed array of the secretproviderclasses.secrets-store.csi.x-k8s.io.

**Comment by jrudley** [comment]
I worked with Microsoft about another issue, but I believe it is directly related. When using Azure insights, the windows portion is completely broke and not sure why they are deploying this. The windows insight pod in kube-system is constantly restarting. It is causing hangs on other windows container pod during terminations. The fix is to remove insights and I have 0 problems now.

**Comment by aramase** [comment]
>I worked with Microsoft about another issue, but I believe it is directly related. When using Azure insights, the windows portion is completely broke and not sure why they are deploying this. The windows insight pod in kube-system is constantly restarting. It is causing hangs on other windows container pod during terminations. The fix is to remove insights and I have 0 problems now.

@jrudley Thank you for updating the issue with your findings. Is there an RCA for the issue? 

@hli07 Do you have windows insight pod running in your cluster too?

**Comment by hli07** [comment]
hi @jrudley / @aramase , I don't have Azure insights enabled nor do I see anything suspiciously restarting in the cluster.

**Comment by jrudley** [comment]
@aramase they are pushing the fix across all the regions and hope to have it done by end of week. css case # 120101424001816

**Comment by aramase** [comment]
@hli07 Thank you for the response. In your cluster, you're observing the events in the terminating pod but that's not blocking the pod from being terminated and deleted, right?

**Comment by hli07** [comment]
thats correct @aramase

**Comment by aramase** [comment]
We've released driver `v0.0.17` and AKV provider `0.0.10` - https://github.com/Azure/secrets-store-csi-driver-provider-azure/releases/tag/0.0.10. With `v0.0.17` we have added checks to skip rotation and reconcile for deleting pods.

Closing this issue with the release and also the primary issue for pod stuck in `Terminating` was caused due to windows insights. Please feel free to open a new issue for any further queries.
